File: examples/cpp/cart_pole/train.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <cstring>
#include <iostream>
#include <stdexcept>
#include <yaml-cpp/yaml.h>

// torchfort
#include "torchfort.h"

// environment
#include "env.h"

// useful for error handling
#define CHECK_TORCHFORT(call)                                                                                          \
  do {                                                                                                                 \
    torchfort_result_t err = call;                                                                                     \
    if (TORCHFORT_RESULT_SUCCESS != err) {                                                                             \
      std::ostringstream os;                                                                                           \
      os << "Error in file " << __FILE__ << ", l" << __LINE__ << ": error code " << err;                               \
      throw std::runtime_error(os.str().c_str());                                                                      \
    }                                                                                                                  \
  } while (false)

int main(int argc, char* argv[]) {

  // load config file
  YAML::Node config = YAML::LoadFile("config_sim.yaml");

  int64_t num_episodes = 10000;
  if (config["num_episodes"]) {
    num_episodes = config["num_episodes"].as<int64_t>();
  }

  int64_t max_steps_per_episode = 500;
  if (config["max_steps_per_episode"]) {
    max_steps_per_episode = config["max_steps_per_episode"].as<int64_t>();
  }

  int eval_frequency = 25;
  if (config["eval_frequency"]) {
    eval_frequency = config["eval_frequency"].as<int>();
  }

  int64_t target_steps = 470;
  if (config["target_steps"]) {
    target_steps = config["target_steps"].as<int64_t>();
  }

  bool normalize_state = false;

  std::cout << "Configuration parameters:" << std::endl;
  std::cout << "Number of episodes: " << num_episodes << std::endl;
  std::cout << "Maximum number of steps per episode: " << max_steps_per_episode << std::endl;
  std::cout << "Evaluation freuency: " << eval_frequency << std::endl;

  // instantiate environment
  CartPoleEnv cenv;
  StateVector state_min, state_max;
  std::tie(state_min, state_max) = cenv.getStateBounds();
  float x_mean = 0.5 * (state_max[0] + state_min[0]);
  float x_width = state_max[0] - state_min[0];
  float theta_mean = 0.5 * (state_max[2] + state_min[2]);
  float theta_width = state_max[2] - state_min[2];

  // instantiate torchfort
#if ENABLE_GPU
  CHECK_TORCHFORT(torchfort_rl_off_policy_create_system("td3_system", "config.yaml", 0, TORCHFORT_DEVICE_CPU));
#else
  CHECK_TORCHFORT(
      torchfort_rl_off_policy_create_system("td3_system", "config.yaml", TORCHFORT_DEVICE_CPU, TORCHFORT_DEVICE_CPU));
#endif

  // define variables
  StateVector state, state_new;
  std::array<float, 1> action;
  std::array<int64_t, 2> state_shape_batch = {1, 4};
  std::array<int64_t, 2> action_shape_batch = {1, 1};
  float reward, total_reward;
  bool terminate;

  // allocate cuda arrays
  float *dstate, *dstate_new, *daction, *dreward;
#ifdef ENABLE_GPU
  cudaSetDevice(0);
  cudaMalloc(&dstate, state.size() * sizeof(float));
  cudaMalloc(&dstate_new, state.size() * sizeof(float));
  cudaMalloc(&daction, action.size() * sizeof(float));
#else
  dstate = static_cast<float*>(std::malloc(state.size() * sizeof(float)));
  dstate_new = static_cast<float*>(std::malloc(state.size() * sizeof(float)));
  daction = static_cast<float*>(std::malloc(action.size() * sizeof(float)));
#endif

  int64_t step_total = 0;
  bool is_eval = false;
  bool is_ready = false;
  float ploss, qloss;
  for (int64_t t = 0; t < num_episodes; ++t) {
    // reset the environment and get the initial state:
    state = cenv.reset();

    // normalize state vector
    if (normalize_state) {
      state[0] = (state[0] - x_mean) / x_width;
      state[2] = (state[2] - theta_mean) / theta_width;
    }

    // reset total reward
    total_reward = 0.;

    // test if this is an eval epoch
    is_eval = (t % eval_frequency == 0);
    std::string prefix = (is_eval ? "EVAL::: " : "TRAIN::: ");

    int64_t step;
    for (step = 0; step < max_steps_per_episode; ++step) {
      // increase total step counter
      step_total++;

      // copy data to device
#ifdef ENABLE_GPU
      cudaMemcpy(dstate, state.data(), state.size() * sizeof(float), cudaMemcpyHostToDevice);
#else
      std::memcpy(dstate, state.data(), state.size() * sizeof(float));
#endif

      // state check
      std::cout << prefix + "state: " << state[0] << ", " << state[1] << ", " << state[2] << ", " << state[3]
                << std::endl;

      if (is_eval)
        CHECK_TORCHFORT(torchfort_rl_off_policy_predict("td3_system", dstate, 2, state_shape_batch.data(), daction, 2,
                                                        action_shape_batch.data(), TORCHFORT_FLOAT, 0));
      else
        CHECK_TORCHFORT(torchfort_rl_off_policy_predict_explore("td3_system", dstate, 2, state_shape_batch.data(),
                                                                daction, 2, action_shape_batch.data(), TORCHFORT_FLOAT,
                                                                0));

        // copy data to host
#ifdef ENABLE_GPU
      cudaMemcpy(action.data(), daction, action.size() * sizeof(float), cudaMemcpyDeviceToHost);
#else
      std::memcpy(action.data(), daction, action.size() * sizeof(float));
#endif

      // action check
      std::cout << prefix + "action: " << action[0] << std::endl;

      // perform an environment step
      std::tie(state_new, reward, terminate) = cenv.step(action[0]);
      total_reward += reward;

      // normalize state vector
      if (normalize_state) {
        state_new[0] = (state_new[0] - x_mean) / x_width;
        state_new[2] = (state_new[2] - theta_mean) / theta_width;
      }

      // copy data to device
#ifdef ENABLE_GPU
      cudaMemcpy(dstate_new, state_new.data(), state_new.size() * sizeof(float), cudaMemcpyHostToDevice);
#else
      std::memcpy(dstate_new, state_new.data(), state_new.size() * sizeof(float));
#endif

      // update replay buffer
      CHECK_TORCHFORT(torchfort_rl_off_policy_update_replay_buffer(
          "td3_system", dstate, dstate_new, 1, &state_shape_batch[1], daction, 1, &action_shape_batch[1], &reward,
          terminate, TORCHFORT_FLOAT, 0));

      // check if episode is over
      if (terminate) {
        break;
      }

      // update state
      state = state_new;
    }

    // training step
    CHECK_TORCHFORT(torchfort_rl_off_policy_is_ready("td3_system", is_ready));
    if (is_ready && !is_eval) {
      CHECK_TORCHFORT(torchfort_rl_off_policy_train_step("td3_system", &ploss, &qloss, 0));
    }

    // report reward
    std::cout << prefix + "Number of steps till failure: " << step << std::endl;
    std::cout << prefix + "Accumulated episode reward: " << total_reward << std::endl;
    std::cout << prefix + "Mean episode reward: " << total_reward / float(step) << std::endl;

    // log to wandb
    CHECK_TORCHFORT(torchfort_rl_off_policy_wandb_log_float("td3_system", "episode reward", t, total_reward));
    CHECK_TORCHFORT(torchfort_rl_off_policy_wandb_log_float("td3_system", "mean episode reward", t,
                                                            total_reward / float(max_steps_per_episode)));

    // save checkpoint:
    CHECK_TORCHFORT(torchfort_rl_off_policy_save_checkpoint("td3_system", "./checkpoint"));

    // stop if we survived target_steps:
    if (is_eval && (step >= target_steps)) {
      std::cout << prefix + ": training finished: episode length was " << step << " steps >= " << target_steps
                << std::endl;
      break;
    }
  }

  // clean up
#ifdef ENABLE_GPU
  cudaFree(dstate);
  cudaFree(dstate_new);
  cudaFree(daction);
#else
  std::free(dstate);
  std::free(dstate_new);
  std::free(daction);
#endif

  return 0;
}

File: examples/fortran/graph/train.f90
! SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
! SPDX-License-Identifier: BSD-3-Clause
!
! Redistribution and use in source and binary forms, with or without
! modification, are permitted provided that the following conditions are met:
!
! 1. Redistributions of source code must retain the above copyright notice, this
!    list of conditions and the following disclaimer.
!
! 2. Redistributions in binary form must reproduce the above copyright notice,
!    this list of conditions and the following disclaimer in the documentation
!    and/or other materials provided with the distribution.
!
! 3. Neither the name of the copyright holder nor the names of its
!    contributors may be used to endorse or promote products derived from
!    this software without specific prior written permission.
!
! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

module utils
  use, intrinsic :: iso_fortran_env, only: real32, int64
  implicit none

  contains
    ! Advecting gaussian bump
    subroutine f_xyt(node_data, nodes, node_types, num_nodes, t)
      implicit none
      integer :: num_nodes
      real(real32) :: node_data(1, num_nodes)
      real(real32) :: nodes(2, num_nodes)
      integer(int64) :: node_types(num_nodes)
      real(real32) :: t

      integer :: i
      real(real32) :: a_x, a_y, x, y
      a_x = 1.0
      a_y = 0.0

      !$acc data copyin(nodes, node_types) copyout(node_data)
      !$acc parallel loop default(present)
      do i = 1, num_nodes
        if (node_types(i) == 0) then
          node_data(1, i) = 0.0
        else
          x = nodes(1, i)
          y = nodes(2, i)
          node_data(1, i) = exp(-20.0 * (x - a_x * t)**2) * exp(-20.0 * (y - a_y * t)**2)
        endif
      enddo
      !$acc end data

    end subroutine f_xyt

    subroutine write_node_data(node_data, num_nodes, fname)
      use, intrinsic :: iso_fortran_env, only: real32
      implicit none
      integer :: num_nodes
      real(real32) :: node_data(1, num_nodes)
      character(len=*) :: fname

      integer :: i

      !$acc update host (node_data)
      open(42, file=fname)
      do i = 1, num_nodes
        write(42, "(e12.4)") node_data(1, i)
      enddo
      close(42)

    end subroutine write_node_data
end module utils

program train
  use, intrinsic :: iso_fortran_env, only: real32, real64, int32, int64
  use torchfort
  use utils
  implicit none

  integer :: i, j, k, idx, istat
  integer :: n1, n2
  integer :: num_edges, num_nodes, num_boundary_nodes
  integer(int64), allocatable :: edge_idx(:,:), node_types(:)
  real(real32), allocatable :: node_feats(:,:), node_feats_rollout(:,:)
  real(real32), allocatable :: edge_feats(:,:)
  real(real32), allocatable :: node_labels(:,:)
  real(real32) :: loss_val, mse, val
  real(real32) :: t, dt, max_t
  character(len=7) :: fidx
  character(len=256) :: filename

  integer :: num_triangles
  real(real32), allocatable:: nodes(:,:)
  integer, allocatable:: triangles(:,:)

  type(torchfort_tensor_list) :: inputs, outputs, labels
  type(torchfort_tensor_list) :: extra_loss_args

  character(len=256) :: configfile = "config.yaml"
  integer :: model_device = 0
  character(len=256) :: node_file = "nodes.txt"
  character(len=256) :: connectivity_file = "connectivity.txt"

  ! Read mesh data
  open(42, file=node_file, action='read')
  read(42, *) num_nodes
  allocate(nodes(2, num_nodes), node_types(num_nodes))
  num_boundary_nodes = 0
  do i = 1, num_nodes
    read(42, *) nodes(1, i), nodes(2, i), node_types(i) 
    if (node_types(i) == 0) num_boundary_nodes = num_boundary_nodes + 1
  end do
  close(42)

  open(42, file=connectivity_file, action='read')
  read(42, *) num_triangles
  allocate(triangles(3, num_triangles))
  do i = 1, num_triangles
    read(42, *) triangles(1, i), triangles(2, i), triangles(3, i)
  end do
  close(42)

  ! Generate edge_idx and edge_feats

  ! Collect all edges from connectivity data (bi-directional)
  num_edges = 6 * num_triangles
  allocate(edge_idx(2, num_edges))
  idx = 1
  do i = 1, num_triangles
    do k = 1, 3
      edge_idx(1, idx) = triangles(k, i)
      edge_idx(2, idx) = triangles(mod(k, 3) + 1, i)
      edge_idx(1, idx + 1) = edge_idx(2, idx)
      edge_idx(2, idx + 1) = edge_idx(1, idx)
      idx = idx + 2
    enddo
  enddo

  ! Set up edge features (dx, dy, magnitude)
  allocate(edge_feats(3, num_edges))
  do i = 1, num_edges
    n1 = edge_idx(1, i) + 1 ! connectivity data is zero-indexed, need to increment
    n2 = edge_idx(2, i) + 1
    edge_feats(1, i) = nodes(1, n2) - nodes(1, n1)
    edge_feats(2, i) = nodes(2, n2) - nodes(2, n1)
    edge_feats(3, i) = sqrt(edge_feats(1, i) * edge_feats(1, i) + edge_feats(2, i) * edge_feats(2, i))
  end do

  allocate(node_feats(1, num_nodes))
  allocate(node_labels(1, num_nodes))
  allocate(node_feats_rollout(1, num_nodes))

  !$acc data copyin(edge_idx, edge_feats, node_feats, node_labels, node_types, node_feats_rollout)

  ! Set up tensor lists
  istat = torchfort_tensor_list_create(inputs)
  istat = torchfort_tensor_list_create(labels)
  istat = torchfort_tensor_list_create(outputs)
  istat = torchfort_tensor_list_create(extra_loss_args)

  !$acc host_data use_device(edge_idx, node_feats, edge_feats, node_labels, node_types, node_feats_rollout)
  istat = torchfort_tensor_list_add_tensor(inputs, edge_idx)
  istat = torchfort_tensor_list_add_tensor(inputs, node_feats)
  istat = torchfort_tensor_list_add_tensor(inputs, edge_feats)

  istat = torchfort_tensor_list_add_tensor(labels, node_labels)

  istat = torchfort_tensor_list_add_tensor(outputs, node_feats_rollout)

  istat = torchfort_tensor_list_add_tensor(extra_loss_args, node_types)
  !$acc end host_data

  ! Set up the model
  istat = torchfort_create_model("mymodel", configfile, model_device)
  if (istat /= TORCHFORT_RESULT_SUCCESS) stop

  ! Training
  t = 0.0
  dt = 0.1
  max_t = 5.0
  print*, "begin training..."
  do i = 1, 100000
    call f_xyt(node_feats, nodes, node_types, num_nodes, t)
    call f_xyt(node_labels, nodes, node_types, num_nodes, t + dt)
    istat = torchfort_train_multiarg("mymodel", inputs, labels, loss_val, extra_loss_args)
    t = t + dt
    if (t >= max_t) t = 0.0
  end do

  ! Rollout validation
  print*, "begin rollout validation..."
  call f_xyt(node_feats, nodes, node_types, num_nodes, 0.0)
  t = 0.0
  do i = 1, int(max_t / dt)
    istat = torchfort_inference_multiarg("mymodel", inputs, outputs)
    call f_xyt(node_feats, nodes, node_types, num_nodes, t + dt)

    ! Compute rollout MSE excluding boundary nodes
    mse = 0.0
    !$acc parallel loop reduction(+:mse) default(present)
    do j = 1, num_nodes
      val = 0.0
      if (node_types(j) /= 0) then
        val = (node_feats(1, j) - node_feats_rollout(1, j))**2
      endif
      mse = mse + val
    enddo
    write(6, '(a3,1x,f5.2,1x,a4,2x,e12.4)'), "t:", t + dt, "mse:", mse / (num_nodes - num_boundary_nodes)

    write(fidx,'(i7.7)') i
    ! Write ground truth
    filename = 'reference_'//fidx//'.txt'
    call write_node_data(node_feats, num_nodes, filename)

    ! Update input for next prediction, excluding boundary nodes
    !$acc parallel loop default(present)
    do j = 1, num_nodes
      if (node_types(j) /= 0) then
        node_feats(1, j) = node_feats_rollout(1, j)
      endif
    enddo

    ! Write prediction with preserved boundary nodes
    filename = 'prediction_'//fidx//'.txt'
    call write_node_data(node_feats, num_nodes, filename)

    t = t + dt
  enddo
  !$acc end data

  ! Destroy tensor lists
  istat = torchfort_tensor_list_destroy(inputs)
  istat = torchfort_tensor_list_destroy(labels)
  istat = torchfort_tensor_list_destroy(outputs)
  istat = torchfort_tensor_list_destroy(extra_loss_args)

end program
